transports/quic/src/connection.rs

```diff
@@ -21,407 +21,87 @@
-                        if matches!(
-                            substream.write_state,
-                            WriteState::Open | WriteState::Closing
-                        ) {
-                            substream.write_state = WriteState::Closed;
-                        }
-                        if let Some(waker) = substream.write_waker.take() {
-                            waker.wake();
-                        }
-                        if let Some(waker) = substream.close_waker.take() {
-                            waker.wake();
-                        }
-                    }
-                }
-                quinn_proto::Event::Stream(quinn_proto::StreamEvent::Stopped {
-                    id,
-                    error_code: _,
-                }) => {
-                    if let Some(substream) = inner.substreams.get_mut(&id) {
-                        substream.write_state = WriteState::Stopped;
-                        if let Some(waker) = substream.write_waker.take() {
-                            waker.wake();
-                        }
-                        if let Some(waker) = substream.close_waker.take() {
-                            waker.wake();
-                        }
-                    }
-                }
-                quinn_proto::Event::DatagramReceived
-                | quinn_proto::Event::Stream(quinn_proto::StreamEvent::Available {
-                    dir: quinn_proto::Dir::Uni,
-                })
-                | quinn_proto::Event::Stream(quinn_proto::StreamEvent::Opened {
-                    dir: quinn_proto::Dir::Uni,
-                }) => {
-                    unreachable!("We don't use datagrams or unidirectional streams.")
-                }
-            }
-        }
-        // TODO: If connection migration is enabled (currently disabled) address
-        // change on the connection needs to be handled.
-
-        self.state.lock().poll_connection_waker = Some(cx.waker().clone());
-        Poll::Pending
-    }
+    type Error = quinn::ConnectionError;
 
     fn poll_inbound(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Result<Self::Substream, Self::Error>> {
-        let mut inner = self.state.lock();
-
-        let substream_id = match inner.connection.streams().accept(quinn_proto::Dir::Bi) {
-            Some(id) => {
-                inner.poll_inbound_waker = None;
-                id
-            }
-            None => {
-                inner.poll_inbound_waker = Some(cx.waker().clone());
-                return Poll::Pending;
-            }
-        };
-        inner.substreams.insert(substream_id, Default::default());
-        let substream = Substream::new(substream_id, self.state.clone());
+        let this = self.get_mut();
 
+        let (send, recv) = futures::ready!(this.incoming.poll_unpin(cx))?;
+        let connection = this.connection.clone();
+        this.incoming = Box::pin(async move { connection.accept_bi().await });
```

Comment:
```suggestion
        this.incoming = async move { connection.accept_bi().await }.boxed();
```

Slightly more succinct, YMMV.